exotations/exotica_core_task_maps/src/avoid_look_at_sphere.cpp

```diff
@@ -0,0 +1,186 @@
+{
+AvoidLookAtSphere::AvoidLookAtSphere() = default;
+AvoidLookAtSphere::~AvoidLookAtSphere() = default;
+
+void AvoidLookAtSphere::AssignScene(ScenePtr scene)
+{
+    scene_ = scene;
+    root_frame_name_ = scene_->GetKinematicTree().GetRootFrameName();
+}
+
+void AvoidLookAtSphere::UpdateAsCostWithoutJacobian(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi)
+{
+    for (int i = 0; i < n_objects_; ++i)
+    {
+        double frac = Eigen::Map<Eigen::Vector3d>(kinematics[0].Phi(i).p.data).topRows<2>().squaredNorm() / radii2_(i);
+        if (frac < 1.0)
+        {
+            phi(i) = 1.0 - 2.0 * frac + frac * frac;
+        }
+        else
+        {
+            phi(i) = 0.0;
+        }
+    }
+}
+
+void AvoidLookAtSphere::UpdateAsInequalityConstraintWithoutJacobian(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi)
+{
+    for (int i = 0; i < n_objects_; ++i)
+    {
+        phi(i) = radii2_(i) - Eigen::Map<Eigen::Vector3d>(kinematics[0].Phi(i).p.data).topRows<2>().squaredNorm();
+    }
+}
+
+void AvoidLookAtSphere::UpdateAsCostWithJacobian(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi, Eigen::MatrixXdRef jacobian)
+{
+    for (int i = 0; i < n_objects_; ++i)
+    {
+        Eigen::Vector2d o = Eigen::Map<Eigen::Vector3d>(kinematics[0].Phi(i).p.data).topRows<2>();
+        double frac = o.squaredNorm() / radii2_(i);
+        if (frac < 1.0)
+        {
+            phi(i) = 1.0 - 2.0 * frac + frac * frac;
+
+            for (int j = 0; j < jacobian.cols(); ++j)
+            {
+                jacobian(i, j) = -4.0 * (1.0 - frac) * kinematics[0].jacobian[i].data.topRows<2>().col(j).dot(o) / radii2_(i);
+            }
+        }
+        else
+        {
+            phi(i) = 0.0;
+        }
+    }
+}
+
+void AvoidLookAtSphere::UpdateAsInequalityConstraintWithJacobian(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi, Eigen::MatrixXdRef jacobian)
+{
+    for (int i = 0; i < n_objects_; ++i)
+    {
+        Eigen::Vector2d o = Eigen::Map<Eigen::Vector3d>(kinematics[0].Phi(i).p.data).topRows<2>();
+        double d = o.norm();
+        phi(i) = radii2_(i) - d * d;
+
+        for (int j = 0; j < jacobian.cols(); ++j)
+        {
+            jacobian(i, j) = -2.0 * kinematics[0].jacobian[i].data.topRows<2>().col(j).dot(o);
+        }
+    }
+}
+
+void AvoidLookAtSphere::PublishObjectsAsMarkerArray()
+{
+    const ros::Time t = ros::Time::now();
+    visualization_msgs::MarkerArray ma;
+    for (int i = 0; i < n_objects_; ++i)
+    {
+        visualization_msgs::Marker m;
+        m.header.stamp = t;
+        m.header.frame_id = "exotica/" + root_frame_name_;
+        m.id = i;
+        m.action = visualization_msgs::Marker::ADD;
+        m.type = visualization_msgs::Marker::SPHERE;
+        m.scale.x = diam_(i);
+        m.scale.y = diam_(i);
+        m.scale.z = diam_(i);
+        Eigen::Vector3d o = Eigen::Map<Eigen::Vector3d>(scene_->GetKinematicTree().FK(frames_[i].frame_A_link_name, KDL::Frame(), root_frame_name_, KDL::Frame()).p.data);
```

Comment:
Hmm, do we really need to run FK for this or isn't this part of the Phi?

Reply:
It is true that we could use `Phi` here. However, in that case because the spheres in RViz are not nice and smooth you can see them rotating (since their orientation as (0,0,0,1) is the eff frame and not the world frame). This makes for some unnecessarily confusing visuals. I am in favor of keeping the call to `FK` to avoid this issue. I don't think speed is really a problem here.

Reply:
Sorry, I don't quite follow the argument - if you use x, y, z from Phi and set the orientation to be what you want - what was the problem? Or how does FK to get x, y, z go around it?

Reply:
Cant you use xyz from Phi and hard-code the orientation to 0,0,0,1 to fix this?

Reply:
Would this work?